allennlp/common/tensor.py

```diff
@@ -1,10 +1,85 @@
-from typing import Dict, Union
-
+from typing import Dict, List, Union
 import torch
 from torch.autograd import Variable
 import numpy
 
 
+def get_lengths_from_binary_sequence_mask(mask: torch.ByteTensor):
+    """
+    Compute sequence lengths for each batch element in a tensor using a
+    binary mask.
+
+    Parameters
+    ----------
+    mask : torch.ByteTensor, required.
+        A 2D binary mask of shape (batch_size, sequence_length) to
+        calculate the per-batch sequence lengths from.
+
+    Returns
+    -------
+    A torch.LongTensor of shape (batch_size,) representing the lengths
+    of the sequences in the batch.
+    """
+    # Flip mask to make padded elements equal to one.
+    inverse_mask = mask == 0
+    # Number of padded elements in sequence.
+    num_padded_elements = inverse_mask.sum(1)
+    # Sequence length is max sequence length minus
+    # the number of padded elements.
+    length_indices = mask.size()[1] - num_padded_elements
+    return length_indices.squeeze().long()
+
+
+def sort_batch_by_length(tensor: torch.FloatTensor, sequence_lengths: torch.LongTensor):
+    """
+    Sort a batch first tensor by some specified lengths.
+
+    Parameters
+    ----------
+    tensor : torch.FloatTensor, required.
+        A batch first Pytorch tensor.
+    sequence_lengths : torch.LongTensor, required.
+        A tensor representing the lengths of some dimension of the tensor which
+        we want to sort by.
+
+    Returns
+    -------
+    sorted_tensor : torch.FloatTensor
+        The original tensor sorted along the batch dimension with respect to sequence_lengths.
+    sorted_sequence_lengths : torch.LongTensor
+        The original sequence_lengths sorted by decreasing size.
+    restoration_indices : torch.LongTensor
+        Indices into the sorted_tensor such that ``sorted_tensor[restoration_indices] == original_tensor``
+    """
+    sorted_sequence_lengths, permutation_index = sequence_lengths.sort(0, descending=True)
+    sorted_tensor = tensor[permutation_index]
+    # This is the equivalent of zipping with index, sorting by the original
+    # sequence lengths and returning the now sorted indices.
+    index_range = torch.range(0, len(sequence_lengths) - 1).long()
+    _, reverse_mapping = permutation_index.sort(0, descending=False)
+    restoration_indices = index_range[reverse_mapping]
+    return sorted_tensor, sorted_sequence_lengths, restoration_indices
+
+
+def get_dropout_mask(dropout_probability: float, shape: List[int]):
+    """
+    Parameters
+    ----------
+    dropout_probability: float, Probability of dropping a dimension of the input.
+    shape: Shape of the tensor you are generating a mask for.
+
+    Return
+    ------
+    A torch.FloatTensor consisting of the binary mask scaled by 1/ (1 - dropout_probability).
```

Comment:
"the binary mask"?  Can you give some more detail here?  It's not obvious what computation this method is performing from this docstring.  Maybe an initial sentence at the beginning of the docstring that says something like "Computes an element-wise dropout mask, where each element in a tensor with shape ``shape`` is dropped out with probability ``dropout_probability``."